### PR TITLE
Asynchronous coordinate transform handler

### PR DESCRIPTION
**Pull request by jheijari**
Requires https://github.com/oskariorg/oskari-server/pull/324

Starting a transformation job returns id to a client.
The client can then start watching the job. 
The request will either time out with the job id or return the transformation result.

### CODE REVIEW

File: service-coordtransform/src/main/java/fi/nls/oskari/spring/CoordinateTransformationAsyncController.java [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+    protected static final Logger log = LogFactory.getLogger(CoordinateTransformationAsyncController.class);
+
+    protected static final String RESPONSE_COORDINATES = "resultCoordinates";
+    protected static final String RESPONSE_DIMENSION = "dimension";
+    protected static final String RESPONSE_JOB_ID = "jobId";
+
+    protected static final String FILE_EXT = "txt";
+    protected static final String FILE_TYPE = "text/plain";
+
+    private static long POLLING_TIMEOUT_MS = 45000l;
```
Comment (jampukka) [inline service-coordtransform/src/main/java/fi/nls/oskari/spring/CoordinateTransformationAsyncController.java]: Should this be 45000?

File: service-coordtransform/src/main/java/fi/nls/oskari/spring/CoordinateTransformationAsyncController.java [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        log.error("Couldn't handle action: " + ROUTE,
+                "Message: ", e.getMessage(),
+                ". Parameters: ", params.getRequest().getParameterMap());
+        writeErrorResponse(params.getResponse(), e.getMessage(),  HttpServletResponse.SC_NOT_IMPLEMENTED, e.getOptions());
+    }
+
+    //add .txt if missing
+    private String addFileExt(String name) {
+        int i = name.lastIndexOf('.');
+        if (i < 0 || i + 1 == name.length()) {
```
Comment (jampukka) [inline service-coordtransform/src/main/java/fi/nls/oskari/spring/CoordinateTransformationAsyncController.java]: If last char is dot we still add another dot before the file extension.

Comment (jampukka) [reply service-coordtransform/src/main/java/fi/nls/oskari/spring/CoordinateTransformationAsyncController.java]: Maybe we can simplify this whole method. Check if name ends with .$EXT and if it doesn't then add the .$EXT. But we probably still have to handle the special case when (while) the last char is a dot.

File: service-coordtransform/src/main/java/fi/nls/oskari/spring/CoordinateTransformationAsyncController.java (hunk 1 above)
Comment (jheijari) [reply service-coordtransform/src/main/java/fi/nls/oskari/spring/CoordinateTransformationAsyncController.java]: The last char is lowercase L for long. True, it's not required and could be left out for clarity.

Comment (jampukka) [reply service-coordtransform/src/main/java/fi/nls/oskari/spring/CoordinateTransformationAsyncController.java]: Oops. my bad. I'm more accustomed to `45000L` (although for floats I use lowercase f, weird.)

File: service-coordtransform/src/main/java/fi/nls/oskari/spring/CoordinateTransformationAsyncController.java [hunk 3]
```diff
@@ -0,0 +1,228 @@
...
+                "Message: ", e.getMessage(),
+                ". Parameters: ", params.getRequest().getParameterMap());
+        writeErrorResponse(params.getResponse(), e.getMessage(),  HttpServletResponse.SC_INTERNAL_SERVER_ERROR, null);
+    }
+
+    private void handleActionParamsException(ActionParamsException e, ActionParameters params) {
+        log.error("Couldn't handle action: " + ROUTE,
+                "Message: ", e.getMessage(),
+                ". Parameters: ", params.getRequest().getParameterMap());
+        writeErrorResponse(params.getResponse(), e.getMessage(),  HttpServletResponse.SC_NOT_IMPLEMENTED, e.getOptions());
```
Comment (jampukka) [inline service-coordtransform/src/main/java/fi/nls/oskari/spring/CoordinateTransformationAsyncController.java]: Is SC_NOT_IMPLEMENTED suitable for an ActionParamsExceptions?

File: service-coordtransform/src/main/java/fi/nls/paikkatietoikkuna/coordtransform/CoordTransWorker.java [hunk 4]
```diff
@@ -0,0 +1,130 @@
...
+    private ConcurrentHashMap<String, TransformParams> paramsMap;
+    private ConcurrentHashMap<String, DeferredResult.DeferredResultHandler> handlerMap;
+    private Cache<Object> resultCache;
+    private static final int READ_TIMEOUT_MS = 1000 * 60 * 15;
+    private static final int CONCURRENT_USERS_LIMIT = 100;
+
+    public static final String RESULT_PENDING = "pending";
+
+    public CoordTransWorker() {
+        clientResponseMap = new ConcurrentHashMap<>();
```
Comment (jampukka) [inline service-coordtransform/src/main/java/fi/nls/paikkatietoikkuna/coordtransform/CoordTransWorker.java]: Is it more or less work to manage 3 different ConcurrentHashMaps? Could it be better to wrap the clientResponse, params and the handler into a single mutable object and only have one map?

Comment (jheijari) [reply service-coordtransform/src/main/java/fi/nls/paikkatietoikkuna/coordtransform/CoordTransWorker.java]: Sounds reasonable, I'll refactor the code.

**Pull request merged**